rts/Sim/Units/Unit.cpp

```diff
@@ -246,6 +246,16 @@ void CUnit::PreInit(const UnitLoadParams& params)
 
 	SetVelocity(params.speed);
 	Move(preFramePos = params.pos.cClampInMap(), false);
+
+	// Force buildings to the 2x2 grid. This normally enforced, but buildings spawned by maps have not
+	// been forced yet.
+	if (unitDef->yardmap.size() != 0) {
+		float x = (std::round(params.pos.x / BUILD_SQUARE_SIZE) * BUILD_SQUARE_SIZE);
+		float z = (std::round(params.pos.z / BUILD_SQUARE_SIZE) * BUILD_SQUARE_SIZE);
```

Comment:
Why not the regular grid?
```suggestion
		float x = (std::round(params.pos.x / SQUARE_SIZE) * SQUARE_SIZE);
		float z = (std::round(params.pos.z / SQUARE_SIZE) * SQUARE_SIZE);
```

Reply:
The purpose here to is find the nearest 2x2 square alignment for the building. As you may recall, buildings need to be aligned on 2x2 squares.

Reply:
Currently they only need to be aligned to the 2x2 game square (16x16 elmo) grid if built by a constructor.

When created via Lua they currently don't need to be aligned at all, which is what this PR will change. I think aligning Lua-spawned buildings to the 1x1 game square (8x8 elmo) grid would be fine.

I think rather than aligning Lua spawned buildings to 2x2 it would be better to eventually provide the option to align constructors to NxN including 1x1.

Reply:
I was hoping to avoid that. The ExitOnly map reduces memory usage, optimize cache utilization, and performance, by using the 2x2 square alignment.

Reply:
> are we really losing anything on the alignment granularity?

The ability to have 1x1 gaps, or to have (say) 3x3 builder alignment. These are quite unimportant to have in TA style games but blizzcraft/aoe style games would want this level of control (gap size is important for melee units, builder alignment is a way to get the proper scale).

> I mean are games relying on 1x1 alignment for buildings for anything? Is Zero-K? BAR certainly doesn't

No but it's begging the question. Games (both existing and not) are largely prevented from relying on it because builders are arbitrarily aligned to 2x2 for no reason.

> in fact BAR would benefit from the enforcement (due to how upgrading mexes work for exmaple.)

```lua
local originalCreateUnit = Spring.CreateUnit
Spring.CreateUnit = function (unitDefID, x, y, z, ...)
  x = math.floor(x / Game.buildSquareSize) * Game.buildSquareSize
  z = math.floor(z / Game.buildSquareSize) * Game.buildSquareSize
  return originalCreateUnit(unitDefID, x, y, z, ...)
end
```

> A misaligned building means you can't build something flush next to it with a builder unit.

Only if the builder alignment isn't also 1x1.

Reply:
Mod rule added.